// bundlemon/lib/main/utils/ci/providers/circleCI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
// https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
const provider = {
    isItMe: (0, utils_1.getEnvVar)('CIRCLECI') === 'true',
    getVars: () => {
        var _a;
        return ({
            ci: true,
            provider: 'circleci',
            owner: (0, utils_1.getEnvVar)('CIRCLE_PROJECT_USERNAME'),
            repo: (0, utils_1.getEnvVar)('CIRCLE_PROJECT_REPONAME'),
            branch: (0, utils_1.getEnvVar)('CIRCLE_BRANCH'),
            commitSha: (0, utils_1.getEnvVar)('CIRCLE_SHA1'),
            // target branch not available in CircleCI
            // https://ideas.circleci.com/cloud-feature-requests/p/provide-env-variable-for-branch-name-targeted-by-pull-request
            // use CI_TARGET_BRANCH to override
            targetBranch: undefined,
            prNumber: (_a = (0, utils_1.getEnvVar)('CIRCLE_PULL_REQUEST')) === null || _a === void 0 ? void 0 : _a.split('/').pop(),
        });
    },
};
exports.default = provider;
