// bundlemon/lib/main/utils/ci/providers/github.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
// https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables
const provider = {
    isItMe: !!(0, utils_1.getEnvVar)('GITHUB_ACTION'),
    getVars: () => {
        var _a, _b;
        const fullRepoName = (0, utils_1.getEnvVar)('GITHUB_REPOSITORY');
        const [owner, repo] = (_a = fullRepoName === null || fullRepoName === void 0 ? void 0 : fullRepoName.split('/')) !== null && _a !== void 0 ? _a : [undefined, undefined];
        const event = (0, utils_1.getEnvVar)('GITHUB_EVENT_NAME');
        const isPr = event === 'pull_request';
        const ref = (_b = (0, utils_1.getEnvVar)('GITHUB_REF')) === null || _b === void 0 ? void 0 : _b.split('/');
        return {
            ci: true,
            provider: 'github',
            owner,
            repo,
            branch: isPr ? (0, utils_1.getEnvVar)('GITHUB_HEAD_REF') : ref === null || ref === void 0 ? void 0 : ref.slice(2).join('/'),
            commitSha: (0, utils_1.getEnvVar)('GITHUB_SHA'),
            targetBranch: (0, utils_1.getEnvVar)('GITHUB_BASE_REF'),
            prNumber: isPr ? ref === null || ref === void 0 ? void 0 : ref[2] : undefined,
            buildId: (0, utils_1.getEnvVar)('GITHUB_RUN_ID'),
        };
    },
};
exports.default = provider;
