// bundlemon/lib/main/utils/ci/providers/travis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
// https://docs.travis-ci.com/user/environment-variables#default-environment-variables
const provider = {
    isItMe: (0, utils_1.getEnvVar)('TRAVIS') === 'true',
    getVars: () => {
        var _a;
        const fullRepoName = (0, utils_1.getEnvVar)('TRAVIS_REPO_SLUG');
        const [owner, repo] = (_a = fullRepoName === null || fullRepoName === void 0 ? void 0 : fullRepoName.split('/')) !== null && _a !== void 0 ? _a : [undefined, undefined];
        const isPushEvent = (0, utils_1.getEnvVar)('TRAVIS_EVENT_TYPE') === 'push';
        const prNumber = (0, utils_1.getEnvVar)('TRAVIS_PULL_REQUEST');
        return {
            ci: true,
            provider: 'travis',
            owner,
            repo,
            branch: isPushEvent ? (0, utils_1.getEnvVar)('TRAVIS_BRANCH') : (0, utils_1.getEnvVar)('TRAVIS_PULL_REQUEST_BRANCH'),
            commitSha: (0, utils_1.getEnvVar)('TRAVIS_COMMIT'),
            targetBranch: isPushEvent ? undefined : (0, utils_1.getEnvVar)('TRAVIS_BRANCH'),
            prNumber: prNumber === 'false' ? undefined : prNumber,
            commitMsg: (0, utils_1.getEnvVar)('TRAVIS_COMMIT_MESSAGE'),
        };
    },
};
exports.default = provider;
