// bundlemon-utils/lib/cjs/diffReport/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDiffReport = exports.calcDiffFiles = void 0;
const consts_1 = require("../consts");
const utils_1 = require("./utils");
function calcDiffFiles(currFiles, baseFiles = []) {
    const filesMap = new Map();
    const basefilesMap = new Map();
    let totalStatus = consts_1.Status.Pass;
    const filesNames = new Set();
    const files = [];
    const stats = {
        currBranchSize: 0,
        baseBranchSize: 0,
    };
    currFiles.forEach((f) => {
        filesMap.set(f.path, f);
        filesNames.add(f.path);
    });
    baseFiles.forEach((f) => {
        basefilesMap.set(f.path, f);
        filesNames.add(f.path);
    });
    Array.from(filesNames)
        .sort()
        .forEach((filename) => {
        var _a, _b, _c, _d, _e, _f, _g;
        const currBranchFile = filesMap.get(filename);
        const baseBranchFile = basefilesMap.get(filename);
        const fileDetails = (currBranchFile || baseBranchFile);
        const diffBytes = ((_a = currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.size) !== null && _a !== void 0 ? _a : 0) - ((_b = baseBranchFile === null || baseBranchFile === void 0 ? void 0 : baseBranchFile.size) !== null && _b !== void 0 ? _b : 0);
        const diffPercent = (0, utils_1.getPercentageDiff)((_c = currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.size) !== null && _c !== void 0 ? _c : 0, (_d = baseBranchFile === null || baseBranchFile === void 0 ? void 0 : baseBranchFile.size) !== null && _d !== void 0 ? _d : 0);
        const change = (0, utils_1.calcChange)({
            isExistsInCurrBranch: !!currBranchFile,
            isExistsInBaseBranch: !!baseBranchFile,
            diffBytes,
        });
        const statusObj = (0, utils_1.getStatusObject)({ currBranchFile, change, diffPercent });
        if (statusObj.status === consts_1.Status.Fail) {
            totalStatus = consts_1.Status.Fail;
        }
        const file = Object.assign(Object.assign(Object.assign({}, statusObj), fileDetails), { size: (_e = currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.size) !== null && _e !== void 0 ? _e : 0, diff: {
                change,
                bytes: diffBytes,
                percent: diffPercent,
            } });
        if (currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.maxSize) {
            file.maxSize = currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.maxSize;
        }
        if (currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.maxPercentIncrease) {
            file.maxPercentIncrease = currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.maxPercentIncrease;
        }
        files.push(file);
        stats.currBranchSize += (_f = currBranchFile === null || currBranchFile === void 0 ? void 0 : currBranchFile.size) !== null && _f !== void 0 ? _f : 0;
        stats.baseBranchSize += (_g = baseBranchFile === null || baseBranchFile === void 0 ? void 0 : baseBranchFile.size) !== null && _g !== void 0 ? _g : 0;
    });
    const diffBytes = stats.currBranchSize - stats.baseBranchSize;
    const diffPercent = (0, utils_1.getPercentageDiff)(stats.currBranchSize, stats.baseBranchSize);
    return {
        files,
        stats: Object.assign(Object.assign({}, stats), { diff: {
                bytes: diffBytes,
                percent: diffPercent,
            } }),
        status: totalStatus,
    };
}
exports.calcDiffFiles = calcDiffFiles;
function generateDiffReport(currInput, baseInput) {
    const filesDiffReport = calcDiffFiles(currInput.files, baseInput === null || baseInput === void 0 ? void 0 : baseInput.files);
    const groupsDiffReport = calcDiffFiles(currInput.groups, baseInput === null || baseInput === void 0 ? void 0 : baseInput.groups);
    const status = filesDiffReport.status === consts_1.Status.Fail || groupsDiffReport.status === consts_1.Status.Fail ? consts_1.Status.Fail : consts_1.Status.Pass;
    return { files: filesDiffReport.files, groups: groupsDiffReport.files, stats: filesDiffReport.stats, status };
}
exports.generateDiffReport = generateDiffReport;
