// bundlemon-utils/lib/esm/consts.js
export var Compression;
(function (Compression) {
    Compression["None"] = "none";
    Compression["Gzip"] = "gzip";
    Compression["Brotli"] = "brotli";
})(Compression || (Compression = {}));
export var DiffChange;
(function (DiffChange) {
    DiffChange["NoChange"] = "No change";
    DiffChange["Update"] = "Update";
    DiffChange["Add"] = "Add";
    DiffChange["Remove"] = "Remove";
})(DiffChange || (DiffChange = {}));
export var Status;
(function (Status) {
    Status["Pass"] = "Pass";
    Status["Fail"] = "Fail";
})(Status || (Status = {}));
export var FailReason;
(function (FailReason) {
    FailReason["MaxSize"] = "MaxSize";
    FailReason["MaxPercentIncrease"] = "MaxPercentIncrease";
})(FailReason || (FailReason = {}));
export var ProjectProvider;
(function (ProjectProvider) {
    ProjectProvider["GitHub"] = "github";
})(ProjectProvider || (ProjectProvider = {}));
export var CommitRecordReviewResolution;
(function (CommitRecordReviewResolution) {
    CommitRecordReviewResolution["Approved"] = "approved";
    CommitRecordReviewResolution["Rejected"] = "rejected";
    CommitRecordReviewResolution["Reset"] = "reset";
})(CommitRecordReviewResolution || (CommitRecordReviewResolution = {}));
