// bundlemon-utils/lib/cjs/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommitRecordReviewResolution = exports.ProjectProvider = exports.FailReason = exports.Status = exports.DiffChange = exports.Compression = void 0;
var Compression;
(function (Compression) {
    Compression["None"] = "none";
    Compression["Gzip"] = "gzip";
    Compression["Brotli"] = "brotli";
})(Compression = exports.Compression || (exports.Compression = {}));
var DiffChange;
(function (DiffChange) {
    DiffChange["NoChange"] = "No change";
    DiffChange["Update"] = "Update";
    DiffChange["Add"] = "Add";
    DiffChange["Remove"] = "Remove";
})(DiffChange = exports.DiffChange || (exports.DiffChange = {}));
var Status;
(function (Status) {
    Status["Pass"] = "Pass";
    Status["Fail"] = "Fail";
})(Status = exports.Status || (exports.Status = {}));
var FailReason;
(function (FailReason) {
    FailReason["MaxSize"] = "MaxSize";
    FailReason["MaxPercentIncrease"] = "MaxPercentIncrease";
})(FailReason = exports.FailReason || (exports.FailReason = {}));
var ProjectProvider;
(function (ProjectProvider) {
    ProjectProvider["GitHub"] = "github";
})(ProjectProvider = exports.ProjectProvider || (exports.ProjectProvider = {}));
var CommitRecordReviewResolution;
(function (CommitRecordReviewResolution) {
    CommitRecordReviewResolution["Approved"] = "approved";
    CommitRecordReviewResolution["Rejected"] = "rejected";
    CommitRecordReviewResolution["Reset"] = "reset";
})(CommitRecordReviewResolution = exports.CommitRecordReviewResolution || (exports.CommitRecordReviewResolution = {}));
